mobilitypath_visualizer/src/mobilitypath_visualizer.cpp

```diff
@@ -0,0 +1,353 @@
+ */
+
+#include <ros/ros.h>
+#include <limits>
+#include <math.h>
+#include "mobilitypath_visualizer.h"
+
+namespace mobilitypath_visualizer {
+
+    namespace
+    {
+        // Helper function for computing 2d distance
+        double compute_2d_distance(const geometry_msgs::Point& pt1, const geometry_msgs::Point&  pt2)
+        {
+            double dx = pt2.x - pt1.x;
+            double dy = pt2.y - pt1.y;
+            return sqrt(dx * dx + dy * dy);
+        }
+    }
+
+    MobilityPathVisualizer::MobilityPathVisualizer() {}
+    void MobilityPathVisualizer::run()
+    {
+        initialize();
+        ros::CARMANodeHandle::setSpinRate(spin_rate_);
+        ros::CARMANodeHandle::setSpinCallback(std::bind(&MobilityPathVisualizer::spinCallback, this));
+        ros::CARMANodeHandle::spin();
+    }
+
+    void MobilityPathVisualizer::initialize()
+    {
+        // init CARMANodeHandle
+        nh_.reset(new ros::CARMANodeHandle());
+        pnh_.reset(new ros::CARMANodeHandle("~"));
+        pnh_->param<double>("spin_rate", spin_rate_, 10.0);
+        nh_->getParam("/vehicle_id", host_id_);
+
+        // init publishers
+        host_marker_pub_ = nh_->advertise<visualization_msgs::MarkerArray>("host_marker", 1);
+        cav_marker_pub_ = nh_->advertise<visualization_msgs::MarkerArray>("cav_marker", 1);
+        label_marker_pub_ = nh_->advertise<visualization_msgs::MarkerArray>("label_marker", 1);
+
+        // init subscribers
+        host_mob_path_sub_ = nh_->subscribe("mobility_path_msg", 1, &MobilityPathVisualizer::callbackMobilityPath, this);
+        cav_mob_path_sub_ = nh_->subscribe("incoming_mobility_path", 1, &MobilityPathVisualizer::callbackMobilityPath, this);
+        tf2_listener_.reset(new tf2_ros::TransformListener(tf2_buffer_));
+    }
+    
+    void MobilityPathVisualizer::callbackMobilityPath(const cav_msgs::MobilityPath& msg)
+    {
+        ROS_ERROR_STREAM("Received new message");
+        if (msg.header.timestamp == 0) //if empty
+            return;
+        ROS_ERROR_STREAM("Received msg from sender: " << msg.header.sender_id << ", and plan id:" << msg.header.plan_id << ", for receiver:" 
+                            << msg.header.recipient_id << ", time:" << msg.header.timestamp << ", at now: " << std::to_string(ros::Time::now().toSec()));
+        if (msg.header.recipient_id.compare(host_id_) != 0 && msg.header.sender_id.compare(host_id_) != 0)
+        {
+            return;
+        }
+        if (msg.header.plan_id == latest_cav_mob_path_msg_[msg.header.sender_id].header.plan_id &&
+            msg.header.timestamp == latest_cav_mob_path_msg_[msg.header.sender_id].header.timestamp)
+        {
+            ROS_ERROR_STREAM("Already received this plan id:" << msg.header.plan_id );
+            return;
+        }
+        latest_cav_mob_path_msg_[msg.header.sender_id] = msg;
+        try
+        {
+            //tf2::convert(tf2_buffer_.lookupTransform("map", "map", ros::Time(0)).transform, map_in_earth_); 
```

Comment:
Since the map to earth transform exists on the TF tree, I think you could avoid this lookup operation entirely and just publish you markers with header.frame_id = "earth"
Rviz should handle the rest